src/bin/cql-stress-cassandra-stress/settings/param/parser.rs

```diff
@@ -0,0 +1,251 @@
+            .iter()
+            .any(|param| param.borrow().required() && !param.borrow().supplied_by_user())
+        {
+            return false;
+        }
+
+        // The group is satisfied - it means that the parameters of this group
+        // were successfully parsed and will be returned to the user.
+        for param in self.params.iter() {
+            param.borrow_mut().set_parsed();
+        }
+        true
+    }
+
+    fn print_help(&self) {
+        let params_size = self.params.len();
+        for (i, param) in self.params.iter().enumerate() {
+            param.borrow().print_usage();
+            if i < params_size - 1 {
+                print!(" ");
+            }
+        }
+    }
+}
+
+/// Parser lets the user define the parameters (see trait [`Param`]).
+/// The parser registers such parameter internally and returns the handle to the user.
+/// Once the user calls `Parser::parse` (which consumes the parser), the values of
+/// the parsed parameters can be retrieved using previously created handles.
+pub struct ParamsParser {
+    command_name: String,
+    params: Vec<ParamCell>,
+    groups: Vec<ParamsGroup>,
+}
+
+impl ParamsParser {
+    pub fn new(command_name: &str) -> Self {
+        Self {
+            command_name: command_name.to_owned(),
+            params: Vec::new(),
+            groups: Vec::new(),
+        }
+    }
+
+    /// Registers the simple parameter provided by the user and returns the handle.
+    /// `value_pattern` has to be a regular expression, otherwise we panic.
+    pub fn simple_param(
+        &mut self,
+        prefix: &'static str,
+        value_pattern: &'static str,
+        default: Option<&'static str>,
+        desc: &'static str,
+        required: bool,
+    ) -> SimpleParamHandle {
+        let param = Rc::new(RefCell::new(SimpleParam::new(
+            prefix,
+            value_pattern,
+            default,
+            desc,
+            required,
+        )));
+
+        self.params.push(Rc::clone(&param) as ParamCell);
+        SimpleParamHandle::new(param)
+    }
+
+    /// Creates a new group of the parameters.
+    pub fn group(&mut self, params: Vec<&dyn ParamHandle>) {
+        self.groups.push(ParamsGroup::new(
+            params.into_iter().map(|handle| handle.cell()).collect(),
+        ))
+    }
+
+    fn parse_groups(&mut self, args: Vec<&str>) -> Result<()> {
+        for arg in args {
+            let mut consumed = false;
+            for group in self.groups.iter_mut() {
+                match group.parse(arg) {
+                    ParamParseResult::Success => {
+                        consumed = true;
+                        break;
+                    }
+                    ParamParseResult::Error(e) => {
+                        return Err(e);
+                    }
+                    _ => (),
+                }
+            }
+            anyhow::ensure!(consumed, "Invalid parameter {}", arg);
+        }
+
+        Ok(())
+    }
```

Comment:
What if all arguments of one group are matched, but there are some additional arguments that matched to different groups? Shouldn't we reject that?

Reply:
And we indeed do reject that. It may not be clear, but `ParamsGroup::satisfied` guarantees that. As stated in the comment:
```
The group is satisfied \iff all of the following requirements are satisfied:
     - number of consumed parameters by the group is equal to the number of CLI arguments
     - every required parameter has been provided by the user
```
The first bullet ensures that there were no parameters that weren't consumed by the group. Now that I look at this comment `the number of CLI arguments` part is not very clear. Obviously, I meant the number of parameters that correspond to the command/option that is currently being parsed.

Reply:
Fixed.
Got rid of this function. I realised it's not necessary to delegate the parsing to the groups. The parser holds the parameters' references as well so he can pretty much do it on his own. The groups are only now used to check for their "satisfaction".